### PR TITLE
Fix parsing GROMACS dihedrals

### PR DESCRIPTION
**Pull request by mattwthompson**
### Description

#979 / #978

### Checklist

- [ ] Add tests
- [ ] Lint
- [ ] Update docstrings

### CODE REVIEW

File: openff/interchange/_tests/unit_tests/interop/gromacs/export/test_export.py [hunk 1]
```diff
@@ -414,6 +418,84 @@ def test_exisiting_mol0_names_overwritten(self, name, sage, ethanol, cyclohexane
...
+                "Electrostatics": 0.05 * unit.kilojoule / unit.mol,
+            },
+        )
+        try:
+            get_gromacs_energies(combined).compare(
+                get_gromacs_energies(combined_from_converted),
+                tolerances={
+                    "Electrostatics": 0.05 * unit.kilojoule / unit.mol,
+                },
+            )
```
Comment (mattwthompson) [inline openff/interchange/_tests/unit_tests/interop/gromacs/export/test_export.py:495]: This is failing pretty catastrophically, not yet sure why

Comment (pbuslaev) [reply openff/interchange/_tests/unit_tests/interop/gromacs/export/test_export.py:495]: Also this is failing

```python
get_openmm_energies(converted_ethanol.combine(converted_cyclohexane)).compare(get_openmm_energies(converted_cyclohexane.combine(converted_ethanol)), tolerances={"Nonbonded": Quantity(0.01, "kilojoule/mole")})
```

**Comment by pbuslaev** [comment]
Ah, ok. I see what is going on. Created `PotentialKey` for different molecules from gromacs are also identical. Thus, only one value is loaded from the [dictionary](https://github.com/openforcefield/openff-interchange/blob/da76d19bbbed3c3d9e8717f5d47a0cd39269076b/openff/interchange/operations/_combine.py#L99). I will work on the potential solution

**Comment by mattwthompson** [comment]
Wow, what a mess. Maybe at import, these should be keyed by something like `f"{molecule_name}-{atomtype1}-{atomtype2}"`?

**Comment by pbuslaev** [comment]
This can be quite a big change, since it has to be introduced into `Bonds`, `Angles`, etc.. I am now testing a different solution, where I am checking if `PotentialKey` is built from indices.

**Comment by mattwthompson** [comment]
Big changes to experimental features are allowed - I consider the state after `Interchange.from_gromacs` to be API- and behavior-unstable and sweeping changes for purposes of fixing bugs or improving performance are okay. By contrast, I'd be apprehensive about changing the results of a vanilla `ForceField.create_interchange` state

Building `PotentialKey` from indices conflicts with the benefit of de-duplicating parameters; this doesn't make much of a difference for these small systems but should for larger systems, proteins, etc.

**Comment by pbuslaev** [comment]
At the moment all my changes are within `_combine` method of `interchange.operations._combine`. What I am trying to do is to check, if the potential key is from indices:

```python
def _check_pot_key_is_by_index(pot_key) -> bool:
        """
        Checks if pot_key is created by index, not by SMARTS.

        Returns
        -------
        True, if generated by index.
        """
        for _atom in pot_key.id.split("-"):
            try:
                int(_atom)
            except ValueError:
                return False
        return True
```

**Comment by pbuslaev** [comment]
But it also won't work with `mol_name`. Currently this scheme is used for `vdW` Handler, and `vdW` are also failing. In the test example, the `mol_name` in the interchanges is the same, so again we fall into same names for different molecules.

**Comment by pbuslaev** [comment]
You are right, but we are dealing with `combine` method. So if I create interchanges like this:
```python
m1 = Molecule("CCC")
m2 = Molecule("CCO")
# skipping conformer generation, etc
i1 = ff.create_interchange(m1.to_topology())
i2 = ff.create_interchange(n2.to_topology())
i1.to_gromacs("m1")
i2.to_gromacs("m2")
i1_c = Interchange.from_gromacs("m1.top", "m1.gro")
i2_c = Interchange.from_gromacs("m2.top", "m2.gro")
```
both `i1_c` and `i2_c` will have handlers associated with `MOL0`. So if I now will try to combine `i1_c.combine(i2_c)` we will have overlap in `PotentialKey` names.

**Comment by mattwthompson** [comment]
I see, that's kinda nasty. This logic might need to get propagated out to `Interchange.combine`: https://github.com/openforcefield/openff-interchange/blob/da76d19bbbed3c3d9e8717f5d47a0cd39269076b/openff/interchange/smirnoff/_gromacs.py#L120-L129

But doing _all_ of the housekeeping for all of the interactions sounds like a pain. I wonder if there's a way to dynamically update all of those keys, or just have them defined as pointers to the molecule name ...

**Comment by mattwthompson** [comment]
There are ways to check if a string is valid SMARTS/SMIRKS, but it'd probably be better to just make a different subclass(es) for these different behaviors ... kinda messy

**Comment by pbuslaev** [comment]
Maybe a simple solution is to add a flag to `PotentialKey` if potentials come from SMIRNOFF or from anywhere else. Then, depending on this flag, different behaviour can be triggered. Maybe `cosmetic_attributes` can be used for it?

**Comment by mattwthompson** [comment]
Yep, we basically do the same thing in a similar situation (storing charges from OpenMM): https://github.com/openforcefield/openff-interchange/blob/da76d19bbbed3c3d9e8717f5d47a0cd39269076b/openff/interchange/interop/openmm/_import/_import.py#L211

File: openff/interchange/_tests/unit_tests/interop/gromacs/export/test_export.py (hunk 1 above)
Comment (mattwthompson) [reply openff/interchange/_tests/unit_tests/interop/gromacs/export/test_export.py]: It's probably H-X bonds that aren't being parsed

**Comment by mattwthompson** [comment]
It's a bit tricky since constraints typically need to know about bonds. Currently this is just dealt with by adding extra logic whenever exporting constraint information

**Review by pbuslaev — APPROVED** [review]

**Pull request merged**